skglm/penalties/separable.py

```diff
@@ -222,42 +223,59 @@ class MCPenalty(BasePenalty):
         "value" = sum_(j=1)^(n_"features") "pen"(abs(w_j))
     """
 
-    def __init__(self, alpha, gamma):
+    def __init__(self, alpha, gamma, positive=False):
         self.alpha = alpha
         self.gamma = gamma
+        self.positive = positive
 
     def get_spec(self):
         spec = (
             ('alpha', float64),
             ('gamma', float64),
+            ('positive', bool_)
         )
         return spec
 
     def params_to_dict(self):
         return dict(alpha=self.alpha,
-                    gamma=self.gamma)
+                    gamma=self.gamma,
+                    positive=self.positive)
 
     def value(self, w):
         return value_MCP(w, self.alpha, self.gamma)
 
     def prox_1d(self, value, stepsize, j):
         """Compute the proximal operator of MCP."""
-        return prox_MCP(value, stepsize, self.alpha, self.gamma)
+        return prox_MCP(value, stepsize, self.alpha, self.gamma, self.positive)
 
     def subdiff_distance(self, w, grad, ws):
         """Compute distance of negative gradient to the subdifferential at w."""
         subdiff_dist = np.zeros_like(grad)
         for idx, j in enumerate(ws):
-            if w[j] == 0:
-                # distance of -grad to alpha * [-1, 1]
-                subdiff_dist[idx] = max(0, np.abs(grad[idx]) - self.alpha)
-            elif np.abs(w[j]) < self.alpha * self.gamma:
-                # distance of -grad_j to (alpha * sign(w[j]) - w[j] / gamma)
-                subdiff_dist[idx] = np.abs(
-                    grad[idx] + self.alpha * np.sign(w[j]) - w[j] / self.gamma)
+            if self.positive:
+                if w[j] < 0:
+                    subdiff_dist[idx] = np.inf
+                elif w[j] == 0:
```

Comment:
duplicated code, no ?
positive just needs to handle w[j] < 0 separately, in that case set subdiff dist to infty and break
```
if positive and w[j] < 0:
    subdiff_dist[idx] = np.inf
    break
and then you keep the old code below

Reply:
Do we agree that the case w[j]  = 0 also needs to be handled separately? As it is currently done for the l1 norm?

Positive l1 https://github.com/scikit-learn-contrib/skglm/blob/562f42b94a1bae58e2c16ad56a0c28f3a9317548/skglm/penalties/separable.py#L44
l1 https://github.com/scikit-learn-contrib/skglm/blob/562f42b94a1bae58e2c16ad56a0c28f3a9317548/skglm/penalties/separable.py#L51

Reply:
I merged the cases `w[j] > 0` for `MCPenalty` and `WeightedMCPenalty`